src/streams/streamsActions.js

```diff
@@ -20,17 +60,27 @@ export const updateExistingStream = (
 
   const auth = getAuth(state);
   const updates = {};
-  if (initialValues.name !== newValues.name) {
-    updates.new_name = maybeEncode(newValues.name);
+  if (changedValues.name !== undefined) {
+    updates.new_name = maybeEncode(changedValues.name);
   }
-  if (initialValues.description !== newValues.description) {
-    updates.description = maybeEncode(newValues.description);
+  if (changedValues.description !== undefined) {
+    updates.description = maybeEncode(changedValues.description);
   }
-  if (initialValues.invite_only !== newValues.invite_only) {
-    updates.is_private = newValues.invite_only;
+  if (changedValues.privacy !== undefined) {
+    const streamProps = privacyToStreamProps(changedValues.privacy);
+
+    // Only send is_web_public if the server will recognize it.
+    // TODO(server-5.0): Remove conditional.
+    if (getZulipFeatureLevel(state) >= 98) {
+      updates.is_web_public = streamProps.is_web_public;
+    }
```

Comment:
Hmm, I think I liked the previous revision's version of this better:
```
    -+    // Only send is_web_public if the server will recognize it. Will it? It
    -+    // should, if we've let the user switch a stream from web-public to not
    -+    // or vice versa.
    -+    // TODO(server-2.1): Remove conditional.
    -+    if (initialPrivacy === 'web-public' || newData.privacy === 'web-public') {
     +      updates.is_web_public = streamProps.is_web_public;
     +    }
```

IIUC, the case where the behavior would differ is where we think the server isn't new enough to recognize the option, but nevertheless for some reason we've allowed the user to say they want to change it. In that case it seems like it's better to go ahead and make the request, and let the server return an error that we'll show the user, rather than silently drop that part of the user's request.

(That and also the converse case: where the server is new enough but the user didn't change to or from web-public, the old logic would leave this property out and the new will include it. That case seems fine either way.)

Reply:
> Hmm, I think I liked the previous revision's version of this better

Ah, I thought you might. 😅

With this function no longer receiving an "initial" `privacy` value—

```
streamsActions [nfc]: Have updateExistingStream only take changed values
```

—it can't identify when the privacy is being switched away from web-public. (It can still tell if it's being switched _to_ web-public, though.) To cover the away-from-web-public case without upsetting old servers, it might be enough to just omit `is_web_public` when switching to any of the three other options? We would still be passing `is_private`, which should [satisfy](https://chat.zulip.org/#narrow/stream/378-api-design/topic/PATCH.20.2Fstreams.2F.7Bstream_id.7D/near/1383984) the requirement I observed that you have to pass at least one of `is_private`, `is_web_public`. I'd want to test manually that things still work. What do you think?

Reply:
> With this function no longer receiving an "initial" `privacy` value—
> 
> ```
> streamsActions [nfc]: Have updateExistingStream only take changed values
> ```
> 
> —it can no longer identify when the privacy is being switched away from web-public. (It can still tell if it's being switched _to_ web-public, though.)

Hmm, I see.



> To cover the away-from-web-public case without upsetting old servers, it might be enough to just omit `is_web_public` when switching to any of the three other options?

That doesn't work if you're switching from web-public to simply public, though, right?

Reply:
OK, given that, this logic is fine.

Basically the concern above means that there's a possible error case we won't catch, and where things silently won't work. But that case only arises if something's already gone wrong that was outside our control: you're on a server that half-supports web-public streams, and I think in fact a server admin has to have gone and manually set a stream web-public in the database (because before this feature-level threshold, there is no way for a client to set it.) Outside our control, and also unlikely -- it's probable that there don't exist any Zulip servers where that situation exists at all.

It's good to catch even situations like that, when we can do so while writing the code in a natural way. But it's OK if we don't always catch them, and it isn't worth contorting the code to handle them.